safekeeper/src/pull_timeline.rs

```diff
@@ -139,6 +193,73 @@ pub async fn stream_snapshot_guts(
     Ok(())
 }
 
+impl Timeline {
+    /// Simple snapshot for an offloaded timeline: we will only upload a renamed partial segment and
+    /// pass a modified control file into the provided tar stream (nothing with data segments on disk, since
+    /// we are offloaded and there aren't any)
+    async fn snapshot_offloaded<W: AsyncWrite + Unpin + Send>(
+        self: &Arc<Timeline>,
+        ar: &mut tokio_tar::Builder<W>,
+        source: NodeId,
+        destination: NodeId,
+    ) -> Result<()> {
+        let shared_state = self.write_shared_state().await;
+
+        let mut control_store = TimelinePersistentState::clone(shared_state.sk.state());
+        // Modify the partial segment of the in-memory copy for the control file to
+        // point to the destination safekeeper.
+        let replace = control_store
+            .partial_backup
+            .replace_uploaded_segment(source, destination)?;
+
+        if let Some(replace) = replace {
+            // The deserialized control file has an uploaded partial. We upload a copy
+            // of it to object storage for the destination safekeeper and send an updated
+            // control file in the snapshot.
+            tracing::info!(
+                "Replacing uploaded partial segment in in-mem control file: {replace:?}"
+            );
+
+            let remote_timeline_path = &self.remote_path;
+            wal_backup::copy_partial_segment(
+                &replace.previous.remote_path(remote_timeline_path),
+                &replace.current.remote_path(remote_timeline_path),
+            )
+            .await?;
+        }
+
+        // Correctness: before sending response, check we didn't get un-offloaded in the background,
+        // i.e. that the object we copied in S3 was not modified in a way that might be inconsistent
+        // with the control file we are sending.
+        match self.try_wal_residence_guard().await {
+            Ok(None) => {
+                // Great, we're still offloaded
+            }
```

Comment:
Unlikely, but what would happen if the timeline was un-offloaded, modified, and then offloaded again in the meanwhile? Can we detect this somehow and at least fail safe, e.g. by comparing the control file modification time?

Reply:
It's not clear to me if a timeline can go through a whole offloaded->restore->offload cycle while this function is still  holding WriteGuardSharedState (or if it can, then can the control file change?)

Will wait for Arseny's input

Reply:
Right, I think it would have to happen before we acquire `WriteGuardSharedState` (e.g. if we end up blocking on it for a long time, or under very high scheduling latency). But you're right -- the control file is accessed under lock, and if we're still offloaded under that lock then I think we should be fine (unless I'm missing some more subtle interaction).

Reply:
With Arseny's help, refined this to what should be a safe behavior that doesn't hog the lock in abfc227a13da01a3b12a45264415e6f4313e89cc -- the logic is now:
1. try_wal_residence_guard
2. acquire WriteGUardSharedState
3. Take a copy of the control file
4. Check if the control file still indicates evicted state: if not, timeline was woken up in the background, bail out so that the API caller can retry.
5. Drop WriteGuardSharedState before doing I/O
6. Attempt S3 copy of partial segments.  If this fails, it indicates timeline was woken up in the background, bail out so that the API caller can retry
7. Send the control file copy from the start.

So now uneviction is explicitly permitted but the result we send should always be consistent because we're sending the control file copy & a copied partial segment that is the one referenced by that copy.  The failure cases should be very rare in practice, and everything that calls our APIs does retries.  It would be nice to improve the error types in future so that we can get more explicit HTTP responses for the "please retry" cases, at the moment most of this API will give 500s for any anyhow::Error.